### ISSUE TITLE
Bogus nullability warning (recent regression)

### ISSUE THREAD

**Issue by jkotas** [dotnet/roslyn#46461]
**Version Used**: 

Microsoft (R) Visual C# Compiler version 3.8.0-2.20379.3 (7238563e)
.NET SDK 5.0.100-rc.1.20380.2 (currently nightly build)

**Steps to Reproduce**:

1. Program.cs:
```
using System;
using System.Linq;
using System.IO;

class Program
{
    private static string GetLatestBuildTools(string androidSdkDir)
    {
        string? buildTools = Directory.GetDirectories(Path.Combine(androidSdkDir, "build-tools"))
            .Select(Path.GetFileName)
            .Where(file => !file!.Contains("-"))
            .Select(file => Version.TryParse(Path.GetFileName(file), out Version? version) ? version : default)
            .OrderByDescending(v => v)
            .FirstOrDefault()?.ToString();

        if (string.IsNullOrEmpty(buildTools))
            throw new ArgumentException($"Android SDK ({androidSdkDir}) doesn't contain build-tools.");

        return buildTools;
    }

    static void Main(string[] args)
    {
    }
}
```
test.csproj:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
    <Nullable>enable</Nullable>
  </PropertyGroup>

</Project>
```
2. `dotnet build`

**Expected Behavior**:

No nullabibility warnings

**Actual Behavior**:

```
C:\repro\Program.cs(12,29): warning CS8603: Possible null reference return. [C:\repro\repro.csproj]
```

**Comment by jkotas** [comment]
This was found during Arcade update validation. https://github.com/dotnet/runtime/pull/40155 is adding workaround.

**Comment by jkotas** [comment]
cc @akoeplinger

**Comment by sharwell** [comment]
Possibly related to #43536. For me, changing `? version : default` to `? version : null` fixed the issue.

**Comment by RikkiGibson** [comment]
More minimal reproducer:

```cs
#nullable enable
using System;

public class C
{
    public void M()
    {
        M1(true, b => b ? "" : default);
    }
    
    public void M1<T,U>(T t, Func<T, U> fn)
    {
        fn(t);
    }
}
```

My current suspicion is that `default` should be causing us to infer a nullable return type for the lambda but it's not doing that for some reason.

**Comment by RikkiGibson** [comment]
This was fixed by #47581, but that doesn't test this precise scenario, so I'd like to PR a regression test.
